diff -Naur dom/media/Makefile.in ../alder/dom/media/Makefile.in
--- dom/media/Makefile.in	2012-09-20 10:50:39.000000000 -0700
+++ ../alder/dom/media/Makefile.in	2012-09-25 19:36:18.000000000 -0700
@@ -19,8 +19,14 @@
 
 include $(topsrcdir)/dom/dom-config.mk
 
+EXTRA_COMPONENTS = \
+  PeerConnection.js \
+  PeerConnection.manifest \
+  $(NULL)
+
 XPIDLSRCS = \
   nsIDOMMediaStream.idl \
+  nsIDOMRTCPeerConnection.idl \
   nsIDOMNavigatorUserMedia.idl \
   $(NULL)
 
@@ -38,6 +44,7 @@
 LOCAL_INCLUDES += \
   -I$(topsrcdir)/media/webrtc/trunk/src \
   $(NULL)
+DIRS += bridge
 endif
 
 include $(topsrcdir)/config/config.mk
diff -Naur dom/media/MediaManager.cpp ../alder/dom/media/MediaManager.cpp
--- dom/media/MediaManager.cpp	2012-09-24 12:36:29.000000000 -0700
+++ ../alder/dom/media/MediaManager.cpp	2012-09-24 13:07:14.000000000 -0700
@@ -23,6 +23,13 @@
 
 namespace mozilla {
 
+// We only support 1 audio and 1 video track for now.
+enum {
+  kVideoTrack = 1,
+  kAudioTrack = 2
+};
+
+
 /**
  * Send an error back to content. The error is the form a string.
  * Do this only on the main thread. The success callback is also passed here
@@ -216,7 +223,16 @@
     NS_ASSERTION(NS_IsMainThread(), "Only call on main thread");
 
     // Create a media stream.
-    nsCOMPtr<nsDOMMediaStream> stream = nsDOMMediaStream::CreateInputStream();
+    nsCOMPtr<nsDOMMediaStream> stream;
+    if (mTrackID == kVideoTrack) {
+      stream = nsDOMMediaStream::CreateInputStream(
+        nsDOMMediaStream::HINT_CONTENTS_VIDEO
+      );
+    } else {
+      stream = nsDOMMediaStream::CreateInputStream(
+        nsDOMMediaStream::HINT_CONTENTS_AUDIO
+      );
+    }
 
     nsPIDOMWindow *window = static_cast<nsPIDOMWindow*>
       (nsGlobalWindow::GetInnerWindowWithId(mWindowID));
@@ -301,12 +317,6 @@
 
   ~GetUserMediaRunnable() {}
 
-  // We only support 1 audio and 1 video track for now.
-  enum {
-    kVideoTrack = 1,
-    kAudioTrack = 2
-  };
-
   NS_IMETHOD
   Run()
   {
diff -Naur dom/media/PeerConnection.js ../alder/dom/media/PeerConnection.js
--- dom/media/PeerConnection.js	1969-12-31 16:00:00.000000000 -0800
+++ ../alder/dom/media/PeerConnection.js	2012-09-25 19:31:14.000000000 -0700
@@ -0,0 +1,476 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+"use strict";
+
+const IDService = {};
+const {classes: Cc, interfaces: Ci, utils: Cu} = Components;
+
+Cu.import("resource://gre/modules/Services.jsm");
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+Cu.import("resource://gre/modules/identity/WebRTC.jsm", IDService);
+
+const PC_CONTRACT = "@mozilla.org/dom/peerconnection;1";
+const PC_ICE_CONTRACT = "@mozilla.org/dom/rtcicecandidate;1";
+const PC_SESSION_CONTRACT = "@mozilla.org/dom/rtcsessiondescription;1";
+
+const PC_CID = Components.ID("{7cb2b368-b1ce-4560-acac-8e0dbda7d3d0}");
+const PC_ICE_CID = Components.ID("{8c5dbd70-2c8e-4ecb-a5ad-2fc919099f01}");
+const PC_SESSION_CID = Components.ID("{5f21ffd9-b73f-4ba0-a685-56b4667aaf1c}");
+
+function IceCandidate(candidate) {
+  this.candidate = candidate;
+  this.sdpMid = null;
+  this.sdpMLineIndex = null;
+}
+IceCandidate.prototype = {
+  classID: PC_ICE_CID,
+
+  classInfo: XPCOMUtils.generateCI({classID: PC_ICE_CID,
+                                    contractID: PC_ICE_CONTRACT,
+                                    classDescription: "IceCandidate",
+                                    interfaces: [
+                                      Ci.nsIDOMRTCIceCandidate,
+                                      Ci.nsIDOMGlobalObjectConstructor
+                                    ],
+                                    flags: Ci.nsIClassInfo.DOM_OBJECT}),
+
+  QueryInterface: XPCOMUtils.generateQI([
+    Ci.nsIDOMRTCIceCandidate, Ci.nsIDOMGlobalObjectConstructor
+  ]),
+
+  constructor: function(win, cand, mid, mline) {
+    this._win = win;
+    this.candidate = cand;
+    this.sdpMid = mid;
+    this.sdpMLineIndex = mline;
+  }
+};
+
+function SessionDescription(type, sdp) {
+  this.type = type;
+  this.sdp = sdp;
+}
+SessionDescription.prototype = {
+  classID: PC_SESSION_CID,
+
+  classInfo: XPCOMUtils.generateCI({classID: PC_SESSION_CID,
+                                    contractID: PC_SESSION_CONTRACT,
+                                    classDescription: "SessionDescription",
+                                    interfaces: [
+                                      Ci.nsIDOMRTCSessionDescription,
+                                      Ci.nsIDOMGlobalObjectConstructor
+                                    ],
+                                    flags: Ci.nsIClassInfo.DOM_OBJECT}),
+
+  QueryInterface: XPCOMUtils.generateQI([
+    Ci.nsIDOMRTCSessionDescription, Ci.nsIDOMGlobalObjectConstructor
+  ]),
+
+  constructor: function(win, type, sdp) {
+    this._win = win;
+    this.type = type;
+    this.sdp = sdp;
+  },
+
+  toString: function() {
+    return JSON.stringify({
+      type: this.type, sdp: this.sdp
+    });
+  }
+};
+
+function PeerConnection() {
+  this._queue = [];
+
+  this._pc = null;
+  this._observer = null;
+  this._identity = null;
+
+  // TODO: Refactor this.
+  this._onCreateOfferSuccess = null;
+  this._onCreateOfferFailure = null;
+  this._onCreateAnswerSuccess = null;
+  this._onCreateAnswerFailure = null;
+  this._onSelectIdentitySuccess = null;
+  this._onSelectIdentityFailure = null;
+  this._onVerifyIdentitySuccess = null;
+  this._onVerifyIdentityFailure = null;
+
+  /**
+   * Everytime we get a request from content, we put it in the queue. If
+   * there are no pending operations though, we will execute it immediately.
+   * In PeerConnectionObserver, whenever we are notified that an operation
+   * has finished, we will check the queue for the next operation and execute
+   * if neccesary. The _pending flag indicates whether an operation is currently
+   * in progress.
+   */
+  this._pending = false;
+
+  // Public attributes.
+  this.onaddstream = null;
+  this.onremovestream = null;
+  this.onicecandidate = null;
+  this.onstatechange = null;
+  this.ongatheringchange = null;
+  this.onicechange = null;
+}
+PeerConnection.prototype = {
+  classID: PC_CID,
+
+  classInfo: XPCOMUtils.generateCI({classID: PC_CID,
+                                    contractID: PC_CONTRACT,
+                                    classDescription: "PeerConnection",
+                                    interfaces: [
+                                      Ci.nsIDOMRTCPeerConnection,
+                                      Ci.nsIDOMGlobalObjectConstructor
+                                    ],
+                                    flags: Ci.nsIClassInfo.DOM_OBJECT}),
+
+  QueryInterface: XPCOMUtils.generateQI([
+    Ci.nsIDOMRTCPeerConnection, Ci.nsIDOMGlobalObjectConstructor
+  ]),
+
+  // Constructor is an explicit function, because of nsIDOMGlobalObjectConstructor.
+  constructor: function(win) {
+    this._pc = Cc["@mozilla.org/peerconnection;1"].
+             createInstance(Ci.IPeerConnection);
+    this._observer = new PeerConnectionObserver(this);
+
+    this._uniqId = Cc["@mozilla.org/uuid-generator;1"]
+                   .getService(Ci.nsIUUIDGenerator)
+                   .generateUUID().toString();
+
+    // Nothing starts until ICE gathering completes.
+    this._queueOrRun({
+      func: this._pc.initialize,
+      args: [this._observer, win, Services.tm.currentThread],
+      wait: true
+    });
+
+    this._win = win;
+    this._winID = this._win.QueryInterface(Ci.nsIInterfaceRequestor)
+                           .getInterface(Ci.nsIDOMWindowUtils).outerWindowID;
+  },
+
+  /**
+   * Add a function to the queue or run it immediately if the queue is empty.
+   * Argument is an object with the func, args and wait properties; wait should
+   * be set to true if the function has a success/error callback that will
+   * call _executeNext, false if it doesn't have a callback.
+   */
+  _queueOrRun: function(obj) {
+    if (!this._pending) {
+      obj.func.apply(this, obj.args);
+      if (obj.wait) {
+        this._pending = true;
+      }
+    } else {
+      this._queue.push(obj);
+    }
+  },
+
+  // Pick the next item from the queue and run it.
+  _executeNext: function() {
+    if (this._queue.length) {
+      let obj = this._queue.shift();
+      obj.func.apply(this, obj.args);
+      if (!obj.wait) {
+        this._executeNext();
+      }
+    } else {
+      this._pending = false;
+    }
+  },
+
+  createOffer: function(onSuccess, onError, constraints) {
+    this._onCreateOfferSuccess = onSuccess;
+    this._onCreateOfferFailure = onError;
+
+    // TODO: Implement constraints/hints.
+    if (!constraints) {
+      constraints = "";
+    }
+
+    this._queueOrRun({
+      func: this._pc.createOffer, args: [constraints], wait: true
+    });
+  },
+
+  createAnswer: function(offer, onSuccess, onError, constraints, provisional) {
+    this._onCreateAnswerSuccess = onSuccess;
+    this._onCreateAnswerFailure = onError;
+
+    if (offer.type != "offer") {
+      if (onError) {
+        onError.onCallback("Invalid type " + offer.type + " passed");
+      }
+      return;
+    }
+
+    if (!offer.sdp) {
+      if (onError) {
+        onError.onCallback("SDP not provided to createAnswer");
+      }
+      return;
+    }
+
+    if (!constraints) {
+      constraints = "";
+    }
+    if (!provisional) {
+      provisional = false;
+    }
+
+    // TODO: Implement provisional answer & constraints.
+    this._queueOrRun({
+      func: this._pc.createAnswer, args: ["", offer.sdp], wait: true
+    });
+  },
+
+  setLocalDescription: function(desc, onSuccess, onError) {
+    this._onSetLocalDescriptionSuccess = onSuccess;
+    this._onSetLocalDescriptionFailure = onError;
+
+    let type;
+    switch (desc.type) {
+      case "offer":
+        type = Ci.IPeerConnection.kActionOffer;
+        break;
+      case "answer":
+        type = Ci.IPeerConnection.kActionAnswer;
+        break;
+      default:
+        if (onError) {
+          onError.onCallback(
+            "Invalid type " + desc.type + " provided to setLocalDescription"
+          );
+          return;
+        }
+        break;
+    }
+
+    this._queueOrRun({
+      func: this._pc.setLocalDescription, args: [type, desc.sdp], wait: true
+    });
+  },
+
+  setRemoteDescription: function(desc, onSuccess, onError) {
+    this._onSetRemoteDescriptionSuccess = onSuccess;
+    this._onSetRemoteDescriptionFailure = onError;
+
+    let type;
+    switch (desc.type) {
+      case "offer":
+        type = Ci.IPeerConnection.kActionOffer;
+        break;
+      case "answer":
+        type = Ci.IPeerConnection.kActionAnswer;
+        break;
+      default:
+        if (onError) {
+          onError.onCallback(
+            "Invalid type " + desc.type + " provided to setLocalDescription"
+          );
+          return;
+        }
+        break;
+    }
+
+    this._queueOrRun({
+      func: this._pc.setRemoteDescription, args: [type, desc.sdp], wait: true
+    });
+  },
+
+  updateIce: function(config, constraints, restart) {
+    return Cr.NS_ERROR_NOT_IMPLEMENTED;
+  },
+
+  addIceCandidate: function(cand) {
+    if (!cand) {
+      throw "Invalid candidate passed to addIceCandidate!";
+    }
+    if (!cand.candidate || !cand.sdpMid || !cand.sdpMLineIndex) {
+      throw "Invalid candidate passed to addIceCandidate!";
+    }
+
+    this._queueOrRun({
+      func: this._pc.addIceCandidate,
+      args: [cand.candidate, cand.sdpMid, cand.sdpMLineIndex], wait: false
+    });
+  },
+
+  addStream: function(stream, constraints) {
+    // TODO: Implement constraints.
+    this._queueOrRun({func: this._pc.addStream, args: [stream], wait: false});
+  },
+
+  removeStream: function(stream) {
+    this._queueOrRun({func: this._pc.removeStream, args: [stream], wait: false});
+  },
+
+  // For testing only.
+  createFakeMediaStream: function(type, muted) {
+    var hint_mute = muted ? 0x80 : 0;
+    if (type == "video") {
+      return this._pc.createFakeMediaStream(Ci.IPeerConnection.kHintVideo |
+                                            hint_mute);
+    }
+    return this._pc.createFakeMediaStream(Ci.IPeerConnection.kHintAudio |
+                                          hint_mute);
+  }
+};
+
+// This is a seperate object because we don't want to expose it to DOM.
+function PeerConnectionObserver(dompc) {
+  this._dompc = dompc;
+}
+PeerConnectionObserver.prototype = {
+  QueryInterface: XPCOMUtils.generateQI([Ci.IPeerConnectionObserver]),
+
+  onCreateOfferSuccess: function(offer) {
+    if (this._dompc._onCreateOfferSuccess) {
+      this._dompc._onCreateOfferSuccess.onCallback({
+        type: "offer", sdp: offer,
+        __exposedProps__: { type: "rw", sdp: "rw" }
+      });
+    }
+    this._dompc._executeNext();
+  },
+
+  onCreateOfferError: function(code) {
+    if (this._dompc._onCreateOfferFailure) {
+      this._dompc._onCreateOfferFailure.onCallback(code);
+    }
+    this._dompc._executeNext();
+  },
+
+  onCreateAnswerSuccess: function(answer) {
+    if (this._dompc._onCreateAnswerSuccess) {
+      this._dompc._onCreateAnswerSuccess.onCallback({
+        type: "answer", sdp: answer,
+        __exposedProps__: { type: "rw", sdp: "rw" }
+      });
+    }
+    this._dompc._executeNext();
+  },
+
+  onCreateAnswerError: function(code) {
+    if (this._dompc._onCreateAnswerFailure) {
+      this._dompc._onCreateAnswerFailure.onCallback(code);
+    }
+    this._dompc._executeNext();
+  },
+
+  onSetLocalDescriptionSuccess: function(code) {
+    if (this._dompc._onSetLocalDescriptionSuccess) {
+      this._dompc._onSetLocalDescriptionSuccess.onCallback(code);
+    }
+    this._dompc._executeNext();
+  },
+
+  onSetRemoteDescriptionSuccess: function(code) {
+    if (this._dompc._onSetRemoteDescriptionSuccess) {
+      this._dompc._onSetRemoteDescriptionSuccess.onCallback(code);
+    }
+    this._dompc._executeNext();
+  },
+
+  onSetLocalDescriptionError: function(code) {
+    if (this._dompc._onSetLocalDescriptionFailure) {
+      this._dompc._onSetLocalDescriptionFailure.onCallback(code);
+    }
+    this._dompc._executeNext();
+  },
+
+  onSetRemoteDescriptionError: function(code) {
+    if (this._dompc._onSetRemoteDescriptionFailure) {
+      this._dompc._onSetRemoteDescriptionFailure.onCallback(code);
+    }
+    this._dompc._executeNext();
+  },
+
+  onStateChange: function(state) {
+    if (state != Ci.IPeerConnectionObserver.kIceState) {
+      return;
+    }
+
+    let iceCb = function() {};
+    let iceGatherCb = function() {};
+    if (this._dompc.onicechange) {
+      iceCb = this._dompc.onicechange;
+    }
+    if (this._dompc.ongatheringchange) {
+      iceGatherCb = this._dompc.ongatheringchange;
+    }
+
+    switch (this._dompc._pc.iceState) {
+      case Ci.IPeerConnection.kIceGathering:
+        iceGatherCb("gathering");
+        break;
+      case Ci.IPeerConnection.kIceWaiting:
+        iceCb("starting");
+        this._dompc._executeNext();
+        break;
+      case Ci.IPeerConnection.kIceChecking:
+        iceCb("checking");
+        this._dompc._executeNext();
+        break;
+      case Ci.IPeerConnection.kIceConnected:
+        // ICE gathering complete.
+        iceCb("connected");
+        iceGatherCb("complete");
+        this._dompc._executeNext();
+        break;
+      case Ci.IPeerConnection.kIceFailed:
+        iceCb("failed");
+        break;
+      default:
+        // Unknown state!
+        break;
+    }
+  },
+
+  onAddStream: function(stream, type) {
+    if (this._dompc.onaddstream) {
+      this._dompc.onaddstream.onCallback({
+        stream: stream, type: type,
+        __exposedProps__: { stream: "r", type: "r" }
+      });
+    }
+    this._dompc._executeNext();
+  },
+
+  onRemoveStream: function(stream, type) {
+    if (this._dompc.onremovestream) {
+      this._dompc.onremovestream.onCallback({
+        stream: stream, type: type,
+        __exposedProps__: { stream: "r", type: "r" }
+      });
+    }
+    this._dompc._executeNext();
+  },
+
+  onAddTrack: function() {
+    // Implement bubbling event up to the appropriate MediaStream.
+  },
+
+  onRemoveTrack: function() {
+    // Implement bubbling event up to the appropriate MediaStream.
+  },
+
+  foundIceCandidate: function(cand) {
+    if (this._dompc.onicecandidate) {
+      this._dompc.onicecandidate.onCallback({
+        candidate: cand,
+        __exposedProps__: { candidate: "rw" }
+      });
+    }
+    this._dompc._executeNext();
+  }
+};
+
+let NSGetFactory = XPCOMUtils.generateNSGetFactory(
+  [IceCandidate, SessionDescription, PeerConnection]
+);
diff -Naur dom/media/PeerConnection.manifest ../alder/dom/media/PeerConnection.manifest
--- dom/media/PeerConnection.manifest	1969-12-31 16:00:00.000000000 -0800
+++ ../alder/dom/media/PeerConnection.manifest	2012-09-25 18:32:48.000000000 -0700
@@ -0,0 +1,11 @@
+component {7cb2b368-b1ce-4560-acac-8e0dbda7d3d0} PeerConnection.js
+component {8c5dbd70-2c8e-4ecb-a5ad-2fc919099f01} PeerConnection.js
+component {5f21ffd9-b73f-4ba0-a685-56b4667aaf1c} PeerConnection.js
+
+contract @mozilla.org/dom/peerconnection;1 {7cb2b368-b1ce-4560-acac-8e0dbda7d3d0}
+contract @mozilla.org/dom/rtcicecandidate;1 {8c5dbd70-2c8e-4ecb-a5ad-2fc919099f01}
+contract @mozilla.org/dom/rtcsessiondescription;1 {5f21ffd9-b73f-4ba0-a685-56b4667aaf1c}
+
+category JavaScript-global-constructor mozRTCPeerConnection @mozilla.org/dom/peerconnection;1
+category JavaScript-global-constructor mozRTCIceCandidate @mozilla.org/dom/rtcicecandidate;1
+category JavaScript-global-constructor mozRTCSessionDescription @mozilla.org/dom/rtcsessiondescription;1
diff -Naur dom/media/bridge/IPeerConnection.idl ../alder/dom/media/bridge/IPeerConnection.idl
--- dom/media/bridge/IPeerConnection.idl	1969-12-31 16:00:00.000000000 -0800
+++ ../alder/dom/media/bridge/IPeerConnection.idl	2012-09-25 18:57:44.000000000 -0700
@@ -0,0 +1,107 @@
+#include "nsIThread.idl"
+#include "nsIDOMWindow.idl"
+
+interface nsIDOMDataChannel;
+interface nsIDOMMediaStream;
+interface nsIDOMWindow;
+
+/* Do not confuse with nsIDOMRTCPeerConnection. This interface is purely for
+ * communication between the PeerConnection JS DOM binding and the C++
+ * implementation in SIPCC.
+ *
+ * See media/webrtc/signaling/include/PeerConnectionImpl.h
+ */
+[scriptable, uuid(84efc76f-41d9-496a-9444-2965d179d419)]
+interface IPeerConnectionObserver : nsISupports
+{
+  /* Constants */
+  const long kReadyState = 0x1;
+  const long kIceState = 0x2;
+  const long kSdpState = 0x3;
+  const long kSipccState = 0x4;
+
+  /* JSEP callbacks */
+  void onCreateOfferSuccess(in string offer);
+  void onCreateOfferError(in unsigned long code);
+  void onCreateAnswerSuccess(in string answer);
+  void onCreateAnswerError(in unsigned long code);
+  void onSetLocalDescriptionSuccess(in unsigned long code);
+  void onSetRemoteDescriptionSuccess(in unsigned long code);
+  void onSetLocalDescriptionError(in unsigned long code);
+  void onSetRemoteDescriptionError(in unsigned long code);
+
+  /* Notification of one of several types of state changed */
+  void onStateChange(in unsigned long state);
+
+  /* Changes to MediaStreams */
+  void onAddStream(in nsIDOMMediaStream stream, in string type);
+  void onRemoveStream();
+  void onAddTrack();
+  void onRemoveTrack();
+
+  /* DataChannel observers */
+  void notifyConnection();
+  void notifyClosedConnection();
+  void notifyDataChannel(in nsIDOMDataChannel channel);
+
+  /* When SDP is parsed and a candidate line is found this method is called.
+   * It should hook back into the media transport to notify it of ICE candidates
+   * listed in the SDP PeerConnectionImpl does not parse ICE candidates, just
+   * pulls them out of the SDP.
+   */
+  void foundIceCandidate(in string candidate);
+};
+
+[scriptable, uuid(942366a9-80fe-4cac-ac97-4fbca45bcbff)]
+interface IPeerConnection : nsISupports
+{
+  const unsigned long kHintAudio = 0x00000001;
+  const unsigned long kHintVideo = 0x00000002;
+
+  const long kActionNone = -1;
+  const long kActionOffer = 0;
+  const long kActionAnswer = 1;
+  const long kActionPRAnswer = 2;
+
+  const long kIceGathering = 0;
+  const long kIceWaiting = 1;
+  const long kIceChecking = 2;
+  const long kIceConnected = 3;
+  const long kIceFailed = 4;
+
+  /* Must be called first. Observer events will be dispatched on the thread provided */
+  void initialize(in IPeerConnectionObserver observer, in nsIDOMWindow window,
+                  [optional] in nsIThread thread);
+
+  /* JSEP calls */
+  void createOffer(in string hints);
+  void createAnswer(in string hints, in string offer);
+  void setLocalDescription(in long action, in string sdp);
+  void setRemoteDescription(in long action, in string sdp);
+
+  /* Adds the stream created by GetUserMedia */
+  void addStream(in nsIDOMMediaStream stream);
+  void removeStream(in nsIDOMMediaStream stream);
+  void closeStreams();
+
+  /* As the ICE candidates roll in this one should be called each time
+   * in order to keep the candidate list up-to-date for the next SDP-related
+   * call PeerConnectionImpl does not parse ICE candidates, just sticks them
+   * into the SDP.
+   */
+  void addIceCandidate(in string candidate, in string mid, in unsigned short level);
+
+  /* Puts the SIPCC engine back to 'kIdle', shuts down threads, deletes state */
+  void close();
+
+  /* For testing purposes only! */
+  nsIDOMMediaStream createFakeMediaStream(in unsigned long hint);
+
+  /* Attributes */
+  readonly attribute string localDescription;
+  readonly attribute string remoteDescription;
+
+  readonly attribute unsigned long iceState;
+  readonly attribute unsigned long readyState;
+  readonly attribute unsigned long sipccState;
+};
diff -Naur dom/media/bridge/Makefile.in ../alder/dom/media/bridge/Makefile.in
--- dom/media/bridge/Makefile.in	1969-12-31 16:00:00.000000000 -0800
+++ ../alder/dom/media/bridge/Makefile.in	2012-08-17 13:10:58.000000000 -0700
@@ -0,0 +1,38 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+DEPTH   = ../../..
+topsrcdir = @top_srcdir@
+srcdir    = @srcdir@
+VPATH   = @srcdir@
+
+include $(DEPTH)/config/autoconf.mk
+
+MODULE        = peerconnection
+XPIDL_MODULE  = peerconnection
+LIBRARY_NAME  = peerconnection
+MODULE_NAME   = peerconnection
+
+IS_COMPONENT   = 1
+EXPORT_LIBRARY = 1
+LIBXUL_LIBRARY = 1
+
+XPIDLSRCS = IPeerConnection.idl
+
+CPPSRCS = \
+  MediaModule.cpp \
+  $(NULL)
+
+LOCAL_INCLUDES += \
+  -I$(topsrcdir)/media/mtransport \
+  -I$(topsrcdir)/media/webrtc/signaling/include \
+  -I$(topsrcdir)/media/webrtc/signaling/src/sipcc/include \
+  -I$(topsrcdir)/media/webrtc/signaling/src/peerconnection \
+  -I$(topsrcdir)/media/webrtc/signaling/src/mediapipeline \
+  -I$(topsrcdir)/media/webrtc/signaling/src/media-conduit \
+  -I$(topsrcdir)/media/webrtc/trunk/third_party/libjingle/source/ \
+  -I$(topsrcdir)/media/webrtc/trunk/third_party/libjingle/source/talk \
+  $(NULL)
+
+include $(topsrcdir)/config/rules.mk
diff -Naur dom/media/bridge/MediaModule.cpp ../alder/dom/media/bridge/MediaModule.cpp
--- dom/media/bridge/MediaModule.cpp	1969-12-31 16:00:00.000000000 -0800
+++ ../alder/dom/media/bridge/MediaModule.cpp	2012-08-07 16:34:15.000000000 -0700
@@ -0,0 +1,47 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "mozilla/ModuleUtils.h"
+#include "nsIClassInfoImpl.h"
+
+#ifdef MOZ_WEBRTC
+
+// wat.
+#include "base/linked_ptr.h"
+using namespace talk_base;
+#include "PeerConnectionImpl.h"
+
+#define PEERCONNECTION_CID \
+{0xb93af7a1, 0x3411, 0x44a8, {0xbd, 0x0a, 0x8a, 0xf3, 0xdd, 0xe4, 0xd8, 0xd8}}
+
+#define PEERCONNECTION_CONTRACTID "@mozilla.org/peerconnection;1"
+
+namespace sipcc
+{
+// Factory defined in sipcc::, defines sipcc::PeerConnectionImplConstructor
+NS_GENERIC_FACTORY_CONSTRUCTOR(PeerConnectionImpl)
+}
+
+// Defines kPEERCONNECTION_CID
+NS_DEFINE_NAMED_CID(PEERCONNECTION_CID);
+
+static const mozilla::Module::CIDEntry kCIDs[] = {
+  { &kPEERCONNECTION_CID, false, NULL, sipcc::PeerConnectionImplConstructor },
+  { NULL }
+};
+
+static const mozilla::Module::ContractIDEntry kContracts[] = {
+  { PEERCONNECTION_CONTRACTID, &kPEERCONNECTION_CID },
+  { NULL }
+};
+
+static const mozilla::Module kModule = {
+  mozilla::Module::kVersion,
+  kCIDs,
+  kContracts
+};
+
+NSMODULE_DEFN(peerconnection) = &kModule;
+
+#endif
diff -Naur dom/media/nsIDOMMediaStream.idl ../alder/dom/media/nsIDOMMediaStream.idl
--- dom/media/nsIDOMMediaStream.idl	2012-08-06 19:51:31.000000000 -0700
+++ ../alder/dom/media/nsIDOMMediaStream.idl	2012-08-07 16:34:15.000000000 -0700
@@ -4,7 +4,7 @@
  * You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 #include "nsISupports.idl"
- 
+
 [scriptable, builtinclass, uuid(f37c2871-4cb7-4672-bb28-c2d601f7cc9e)]
 interface nsIDOMMediaStream : nsISupports
 {
diff -Naur dom/media/nsIDOMRTCPeerConnection.idl ../alder/dom/media/nsIDOMRTCPeerConnection.idl
--- dom/media/nsIDOMRTCPeerConnection.idl	1969-12-31 16:00:00.000000000 -0800
+++ ../alder/dom/media/nsIDOMRTCPeerConnection.idl	2012-09-25 18:57:46.000000000 -0700
@@ -0,0 +1,92 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "nsISupports.idl"
+#include "nsIDOMMediaStream.idl"
+
+[scriptable, function, uuid(eb9c563c-3b09-4565-9317-eca96ae0c538)]
+interface RTCPeerConnectionCallback : nsISupports
+{
+  void onCallback(in jsval value);
+};
+
+[scriptable, function, uuid(55546efd-287b-4460-8283-0592875b890f)]
+interface RTCPeerConnectionCallbackVoid : nsISupports
+{
+  void onCallback();
+};
+
+[scriptable, uuid(05d7375e-b024-4951-a570-c6642105ad35)]
+interface nsIDOMRTCSessionDescription : nsISupports
+{
+  attribute DOMString sdp;
+  attribute DOMString type;
+};
+
+[scriptable, uuid(df176474-e20a-4f42-a85b-b0414d634cf0)]
+interface nsIDOMRTCIceCandidate : nsISupports
+{
+  attribute DOMString candidate;
+  attribute DOMString sdpMid;
+  attribute unsigned short sdpMLineIndex;
+};
+
+/* See http://dev.w3.org/2011/webrtc/editor/webrtc.html */
+[scriptable, uuid(94628e70-e96f-4170-871c-f993a49f065a)]
+interface nsIDOMRTCPeerConnection : nsISupports
+{
+  void createOffer(in RTCPeerConnectionCallback successCallback,
+    [optional] in RTCPeerConnectionCallback failureCallback,
+    [optional] in jsval constraints);
+
+  void createAnswer(in nsIDOMRTCSessionDescription offer,
+    in RTCPeerConnectionCallback successCallback,
+    [optional] in RTCPeerConnectionCallback failureCallback,
+    [optional] in jsval constraints,
+    [optional] in bool createProvisionalAnswer);
+
+  void setLocalDescription(in nsIDOMRTCSessionDescription desc,
+    [optional] in RTCPeerConnectionCallback successCallback,
+    [optional] in RTCPeerConnectionCallback failureCallback);
+
+  void setRemoteDescription(in nsIDOMRTCSessionDescription desc,
+    [optional] in RTCPeerConnectionCallback successCallback,
+    [optional] in RTCPeerConnectionCallback failureCallback);
+
+  void updateIce([optional] in jsval configuration,
+    [optional] in jsval constraints,
+    [optional] in bool restart);
+
+  void addIceCandidate(in nsIDOMRTCIceCandidate candidate);
+
+  void addStream(in nsIDOMMediaStream stream,
+    [optional] in jsval constraints);
+
+  void removeStream(in nsIDOMMediaStream stream);
+  void close();
+
+  /* Readonly attributes */
+  readonly attribute DOMString iceState;
+  readonly attribute DOMString iceGatheringState;
+
+  readonly attribute DOMString readyState;
+  readonly attribute jsval localDescription;
+  readonly attribute jsval remoteDescription;
+
+  readonly attribute jsval localStreams; // MediaStream[]
+  readonly attribute jsval remoteStreams; // MediaStream[]
+
+  /* Event handlers. TODO: Use real EventTarget */
+  attribute RTCPeerConnectionCallback onaddstream;
+  attribute RTCPeerConnectionCallback onremovestream;
+  attribute RTCPeerConnectionCallback onicecandidate;
+  attribute RTCPeerConnectionCallback onstatechange;
+  attribute RTCPeerConnectionCallback ongatheringchange;
+  attribute RTCPeerConnectionCallback onicechange;
+
+  /* For purposes only! Remove in production */
+  nsIDOMMediaStream createFakeMediaStream(in string type,
+                                          [optional] in boolean mute
+                                          );
+};
diff -Naur browser/installer/package-manifest.in ../alder/browser/installer/package-manifest.in
--- browser/installer/package-manifest.in 2012-09-19 14:03:12.000000000 -0700
+++ ../alder/browser/installer/package-manifest.in  2012-09-24 11:12:53.000000000 -0700
@@ -266,6 +266,7 @@
 @BINPATH@/components/necko.xpt
 @BINPATH@/components/loginmgr.xpt
 @BINPATH@/components/parentalcontrols.xpt
+@BINPATH@/components/peerconnection.xpt
 @BINPATH@/components/places.xpt
 @BINPATH@/components/plugin.xpt
 @BINPATH@/components/pref.xpt
@@ -491,6 +492,9 @@
 @BINPATH@/components/TCPSocket.js
 @BINPATH@/components/TCPSocket.manifest
 
+@BINPATH@/components/PeerConnection.js
+@BINPATH@/components/PeerConnection.manifest
+
 #ifdef ENABLE_MARIONETTE
 @BINPATH@/chrome/marionette@JAREXT@
 @BINPATH@/chrome/marionette.manifest
diff -Naur toolkit/library/Makefile.in ../alder/toolkit/library/Makefile.in
--- toolkit/library/Makefile.in 2012-09-25 19:32:17.000000000 -0700
+++ ../alder/toolkit/library/Makefile.in  2012-09-24 11:12:56.000000000 -0700
@@ -111,7 +111,7 @@
 endif #}
 
 ifdef MOZ_B2G_BT #{
-STATIC_LIBS += mozdbus_s mozipcunixsocket_s
+STATIC_LIBS += mozdbus_s mozipcsocket_s
 endif #}
 
 ifeq (gonk,$(MOZ_WIDGET_TOOLKIT))
@@ -376,6 +376,26 @@
 
 EXTRA_DSO_LDOPTS += $(call EXPAND_LIBNAME_PATH,gkmedias,$(DIST)/lib)
 
+ifdef MOZ_WEBRTC
+COMPONENT_LIBS += peerconnection
+EXTRA_DSO_LDOPTS += \
+  $(DEPTH)/media/mtransport/build/$(LIB_PREFIX)mtransport.$(LIB_SUFFIX) \
+  $(DEPTH)/media/webrtc/signaling/signaling_ecc/$(LIB_PREFIX)ecc.$(LIB_SUFFIX) \
+  $(DEPTH)/media/webrtc/signaling/signaling_sipcc/$(LIB_PREFIX)sipcc.$(LIB_SUFFIX) \
+  $(NULL)
+ifdef MOZ_WEBRTC_IN_LIBXUL
+include $(topsrcdir)/media/webrtc/shared_libs.mk
+EXTRA_DSO_LDOPTS += $(WEBRTC_LIBS)
+ifeq (WINNT,$(OS_TARGET))
+EXTRA_DSO_LDOPTS += \
+  -LIBPATH:"$(MOZ_DIRECTX_SDK_PATH)/lib/$(MOZ_DIRECTX_SDK_CPU_SUFFIX)" \
+  $(NULL)
+OS_LIBS += $(call EXPAND_LIBNAME,secur32 crypt32 iphlpapi strmiids dmoguids wmcodecdspuuid amstrmid msdmo wininet)
+endif
+endif
+endif
+
+
 ifdef MOZ_SYDNEYAUDIO
 ifeq ($(OS_ARCH),Linux)
 EXTRA_DSO_LDOPTS += $(MOZ_ALSA_LIBS)
diff -Naur toolkit/library/nsStaticXULComponents.cpp ../alder/toolkit/library/nsStaticXULComponents.cpp
--- toolkit/library/nsStaticXULComponents.cpp 2012-09-14 11:44:07.000000000 -0700
+++ ../alder/toolkit/library/nsStaticXULComponents.cpp  2012-09-06 18:02:10.000000000 -0700
@@ -149,6 +149,12 @@
 #define JSCTYPES_MODULE
 #endif
 
+#if defined(MOZ_WEBRTC)
+#define PEERCONNECTION_MODULE MODULE(peerconnection)
+#else
+#define PEERCONNECTION_MODULE
+#endif
+
 #ifndef MOZ_APP_COMPONENT_MODULES
 #if defined(MOZ_APP_COMPONENT_INCLUDE)
 #include MOZ_APP_COMPONENT_INCLUDE
@@ -221,6 +227,7 @@
     MODULE(nsTelemetryModule)                \
     MODULE(jsinspector)                      \
     MODULE(jsdebugger)                       \
+    PEERCONNECTION_MODULE                    \
     /* end of list */
 
 #define MODULE(_name) \
